### ISSUE TITLE
How to apply the `private`, `export` and `public export` accessibility qualifiers?

### ISSUE THREAD

**Issue by nicolabotta** [idris-lang/Idris-dev#2990]
This issue is a follow-up to #2985. The idea is to achieve some shared understanding of how the new `private`, `export` and `public export` accessibility qualifiers can be applied -- possibly in conjunction with `%hide` and `%freeze` directives and with qualified `import public` rules -- to control the accessibility of features of imported modules. I will discuss an example that I think is paradigmatic. Consider the `Foo` module
```
> module Foo

> %default total
> %access public export

> data Foo : Nat -> Type where
>   MkFoo  : {n : Nat} -> Foo (S n)
```
together with its `FooOps` and `FooProps` library modules
```
> module FooOps

> import Foo

> %default total
> %access public export
> -- %access export

> fromSucc : (m : Nat) -> (n : Nat) -> S m = n -> Foo n
> fromSucc m n prf = s2 where
>   s1 : Foo (S m)
>   s1 = MkFoo
>   s2 : Foo n
>   s2 = replace prf s1

> toNat : Foo n -> Nat
> toNat {n} _ = n
```
,
```
> module FooProps

> import Foo
> import FooOps

> %default total
> %access export

> toFromLemma : (m : Nat) -> (n : Nat) -> (p : S m = n) -> toNat (fromSucc m n p) = n 
> toFromLemma m n p = Refl
```
and the application-level program
```
> module Bar

> import Foo
> import FooOps

> %default total

> foo : Foo 1
> foo = fromSucc Z (S Z) Refl
```
The question is: **should `FooOps.fromSucc` and `FooOps.toNat` be qualified as `private`, `export` or `public export`?** Before trying to answer this question, let's observe that:

1) `FooProps` needs to access the implementation of `FooOps.fromSucc` and `FooOps.toNat` for `toFromLemma` to type check.

2) `Bar` does not need to (and, for reasons that I do not want to discuss here, should not) access  the implementation of `FooOps.fromSucc` and `FooOps.toNat`.

Thus, **in absence of other mechanisms** for controlling the accessibility of `FooOps.fromSucc` and `FooOps.toNat` in `FooProps` and `Bar`, **we have to do with a specification that we cannot fulfill: `FooOps.fromSucc` and `FooOps.toNat` have to be qualified as `public export` because of 1) and they have to be qualified as `export`because of 2)!**

Thus, a refinement of the above question is: **Can we find suitable combinations of accessibility qualifiers for `FooOps.fromSucc` and `FooOps.toNat`, import qualifiers and perhaps `%hide` and `%freeze` directives that allow for the implementations of `FooOps.fromSucc` and `FooOps.toNat` to be accessible in `FooProps` but not in `Bar`?**

In #2985, jfdm suggests that the "use of the modifier `public` in an `import` declaration allows one to promote the imported module's definition to the definition of the calling module". I do not understand what this means but replacing
```
> %access public export
``` 
with
```
> %access export
``` 
in `FooOps` (to make the implementations of `FooOps.fromSucc` and `FooOps.toNat` not accessible in `Bar`) and
```
> import FooOps
```
with
```
> import public FooOps
```
in `FooProps` does not make the implementations of `FooOps.fromSucc` and `FooOps.toNat` accessible in `FooProps`:
```
- + Errors (1)
 `-- FooProps.lidr line 10 col 12:
     When checking right hand side of toFromLemma with expected type
             toNat (fromSucc m n p) = n
     
     Type mismatch between
             n = n (Type of Refl)
     and
             toNat (fromSucc m n p) = n (Expected type)
     
     Specifically:
             Type mismatch between
                     n
             and
                     toNat (fromSucc m n p)
```
Let's try the other way round: instead of increasing the accessibility of `FooOps`features in `FooProps`, we restrict it in `Bar`. `FooOps.fromSucc` and `FooOps.toNat` are now declated to be `public export` accessible but their implementations are frozen in `Bar`:
```
> module Bar

> import Foo
> import FooOps

> %default total
> %freeze FooOps.fromSucc
> %freeze FooOps.toNat

> foo : Foo 1
> foo = fromSucc Z (S Z) Refl

> toFromLemma' : (m : Nat) -> (n : Nat) -> (p : S m = n) -> toNat (fromSucc m n p) = n 
> toFromLemma' m n p = Refl
```
This works (`toFromLemma'` does not type check) but raises the questions of **whether this is a canonical way the accessibility qualifiers are meant to be used** and, if so, why the default qualifier isn't then `public export` instead of `private`: if clients can only restrict (but not increase ) the accessibility of imported features, it would seem natural for the default to be `public export` as it was before 0.10! 

Also, #2977 suggests that `%freeze` directive can behave a bit fuzzily and I am not sure I understand the difference between a
```
> export lala : Nat
> lala = 3
```
declaration and a 
```
> public export lala : Nat
> lala = 3
> %freeze lala
```
declaration.

**Tentative wrap-up:** I think that it is hardly questionable that clients have to be able to decide the accessibility level of imported features (I might be wrong, of course). Under the current rules, it appears that clients can only restrict the accessibility of imported features via `%freeze` and `%hide` directives. Thus, it seems that a suitable default value for accessibility should be `public export`, not `private`. The above analysis also suggests that, without `%freeze` directives, the new accessibility rules are not flexible enough to allow for clients to set the proper accessibility level of imported features. This suggests that `%freeze` is essential for controlling accessibility and should perhaps be promoted to a language keyword like `private`, `export` and `public export`.

**Comment by edwinb** [comment]
I think a problem with the new access rules is that there is too much interaction between the rules for each function. So if a public export function (say 'Foo.foo') uses an imported function (say 'Bar.bar'), but a module which imports Foo doesn't import Bar, you get surprising behaviour because Bar.bar isn't visible.

The internal problem is that 'private' also makes things inaccessible to the type checker and evaluator, meaning that if private things accidentally show up due to some evaluation, a program will no longer type check when it probably should. 

What I want to happen is that 'export' gives you just the name, and 'public export' gives you the entire reduction behaviour. That's not quite what we have at the minute, and the restrictions we use are too painful in any case.

I'm going to try something much simpler, in which 'private' merely makes things inaccessible to the programmer, and in which there are no restrictions on the evaluator under a 'public export'. This is probably also what people expect and reduces the need for excessive freeze/hide directives.

**Comment by nicolabotta** [comment]
Edwin, thanks for looking into this issue! I do not have a concrete suggestion at this point and I am afraid that I have been mixing up two concerns that should probably be kept well separated.

The first concern is that of controlling which **names** of a given module should be visible in client modules (and in client modules of client modules). An aspect of this concern is the question of whether a module should be able to impose restrictions on the visibility of certain module names for client modules (for instance, for implementing abstract data types) or whether the visibility of imported names should be fully controlled by the importing modules (or, perhaps, something in between).  

In any case, what we are talking about here is visibility of names, not of implementations. I would expect mechanisms that control the visibility of names to have a significant impact on how Idris users perceive and apply the Idris module system. I also would expect such mechanisms to be made available through keywords like `private`, `public`, etc. But I would not expect mechanisms for controlling the visibility of names to be motivated by (or to be explained in terms of) how the type checker or the evaluator work. In other words, I think that it would be a bad idea to give an operational meaning to such mechanisms in terms of operations that Idris users should not be primarily concerned with. 

The second concern is that of the visibility of **implementations** the purpose of type checking. Here I would expect users to be able to exercise some form of control, for instance, to make type checking faster or more reliable. But I would expect mechanisms that control the visibility of implementations to be well separated from mechanisms that control the visibility of names. I am possibly mistaken but it seems to me that the question of which implementations should be made available for type checking a given expression has little to do with the usage of a modules system and with the visibility of names in that system. 

In a nutshell: **I think that controlling the visibility of names (for avoiding name clashes, hiding information, implementing ADTs) and controlling the visibility of implementations (for speeding up type checking, exploiting optimization opportunities, inlining, etc.) are rather orthogonal concerns that should be supported by well distinct sets of language keywords on the one hand and optimization directives on the other hand.**
